Clarify when to send ACK_FREQUENCY frame (#275)
quicwg/ack-frequency: draft-ietf-quic-ack-frequency.md

unnecessarily high number of acknowledgments when min_rtt is much smaller than
smoothed_rtt.

Note that the average congestion window and the smoothed RTT change over the lifetime of a
connection and therefore might require sending frequent ACK_FREQUENCY frames to
ensure optimal performance. As both the congestion window and RTT estimates

> change continiously, of course not every change should trigger an update.

Author: ```suggestion
may change continuously, of course not every change should trigger an update.
```

> Usually, it is not nessecary to send an ACK_FREQUENCY frame more than once per
> RTT. Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the

Author: ```suggestion
Usually, it is not nessecary to send an ACK_FREQUENCY frame more than once per
RTT and likely it needs to be sent even less frequently.
Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the
```

> Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the
> average congestion window or RTT estimates is detected that impacts the local

Author: ```suggestion
Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the
longer-term development of the congestion window or RTT estimates is detected that impacts the local
```

Reviewer: I'd prefer something like: "Ideally, an ACK_FREQUENCY frame ought to be sent after a change in the
the window or the smoothed RTT that impacts the local
setting of the reordering threshold or locally-selected calculation of the
Ack-Eliciting Threshold or Requested Max Ack Delay."

Author: ```suggestion
Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the congestion window or RTT estimates is detected that impacts the local
```

> Ideally, an ACK_FREQUENCY frame is sent only if a relevant change

Author: ```suggestion
Ideally, an ACK_FREQUENCY frame is sent only when a relevant change
```

setting of the reordering threshold or locally-selected calculation of the

> Ack-Eliciting Threshold or Requested Max Ack Delay.

Reviewer 2: I wonder if rearranging the paragraph would make things a bit more clear. Taking the proposed text and shuffling things a bit, something like

> Usually, it is sufficient to send an ACK_FREQUENCY frame once per RTT or at a
lower frequency. However, the congestion window and RTT estimate can change over
the lifetime of a connection, which might require sending frames at a higher
frequency in order to achive optimal performance. Not every change to congestion
window or RTT estimate warrants sending an ACK_FREQUENCY frame, ideally sending
occurs only when a change to congestion window or smoothed RTT impacts the local
setting of: the reordering threshold, or locally-selected calculation of the
Ack-Eliciting Threshold, or Requested Max Ack Delay.

Author: thanks for the proposal. I don't think your first sentence is putting the emphasis right. therefore I slightly prefer the current order of statements but I made some small wording updates that hopefully help. I will merge for now but if further editorial improvements are needed we can probably still address this later with another PR. Thanks!

unnecessarily high number of acknowledgments when min_rtt is much smaller than
smoothed_rtt.

Note that the average congestion window and the smoothed RTT change over the lifetime of a

> connection and therefore might require sending frequent ACK_FREQUENCY frames to

Author: I don't think frequent implies one per RTT. I think frequent is more correct because it's about the period you send ACKs and not the number of ACKs.

> connection and therefore might require sending frequent ACK_FREQUENCY frames to

Reviewer: I don't much like "many"or "frequent", is it enough to say "additional"?

> connection and therefore might require sending frequent ACK_FREQUENCY frames to

Author: addition to what?

> connection and therefore might require sending frequent ACK_FREQUENCY frames to

Author: ```suggestion
connection and therefore might require sending updates in an ACK_FREQUENCY frames to
```

ensure optimal performance. As both the congestion window and RTT estimates
may change continuously, of course not every change should trigger an update.

> Usually, it is not nessecary to send an ACK_FREQUENCY frame more than once per

Author: ```suggestion
Usually, it is not necessary to send an ACK_FREQUENCY frame more than once per
```

RTT and likely it needs to be sent even less frequently.
Ideally, an ACK_FREQUENCY frame is sent only if a relevant change in the
average congestion window or RTT estimates is detected that impacts the local

> setting of the reordering threshold or local-selected calculation of the

Author: ```suggestion
setting of the reordering threshold or locally-selected calculation of the
```

> Ack-Eliciting Threshold or Requested Max Ack Delay.

Reviewer 3: There's that "average" again.

The problem I have with this text is that it says that you compute the value and then only ask the peer to change if the computed value changes.  However, this is still going to thrash quite a bit as estimates fluctuate.  Especially during slow start, where big changes are commonplace (but more ACK frames is fine).

I would instead recommend some amount of hysteresis in values.  That should have the desired effect.

Author: The congestion window for most congestion control algorithm makes a sawtooth over a period of cwnd/2 RTTs. The congestion increases by one for cwnd/2 RTT and then is halved when a congestion indication occurs. That means the congestion window changes constantly but the average (over cwnd/2 RTT) stays constant. Only if there is a real change in bandwidth and therefore the average value changes, it makes sense to also adjust the ack frequency.

I tried to change the working to "longer-term development". @martinthomson does that help?

Reviewer: I suspect this may just trying to explain too much, not normally sending one more than a RTT seems reasonable; adjusting if there is a significant change in cwnd or smoothed RTT also seems useful to note. I didn't like "average cwnd" or longer term development much though.

Author: @gorryfair what do you mean by "not normally sending one more than a RTT "?

Do you have any wording suggestion?
